flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/StateChangelogStorageLoader.java

```diff
@@ -18,26 +18,81 @@
+import org.apache.flink.configuration.Configuration;
 import org.apache.flink.core.plugin.PluginManager;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
 import java.util.Iterator;
 import java.util.ServiceLoader;
+import java.util.concurrent.locks.ReentrantLock;
 
 import static org.apache.flink.shaded.guava18.com.google.common.collect.Iterators.concat;
 
 /** A thin wrapper around {@link PluginManager} to load {@link StateChangelogStorage}. */
 @Internal
 public class StateChangelogStorageLoader {
-    private final PluginManager pluginManager;
 
-    public StateChangelogStorageLoader(PluginManager pluginManager) {
-        this.pluginManager = pluginManager;
+    private static final Logger LOG = LoggerFactory.getLogger(StateChangelogStorageLoader.class);
+
+    /** Object used to protect calls to specific methods. */
+    private static final ReentrantLock LOCK = new ReentrantLock(true);
+
+    /**
+     * Mapping of state changelog storage identifier to the corresponding storage factories,
+     * populated in {@link StateChangelogStorageLoader#initialize(PluginManager)}.
+     */
+    private static final HashMap<String, StateChangelogStorageFactory>
+            STATE_CHANGELOG_STORAGE_FACTORIES = new HashMap<>();
+
+    static {
+        // Guarantee to trigger once.
+        initialize(null);
+    }
+
+    public static void initialize(PluginManager pluginManager) {
+        LOCK.lock();
+        try {
+            STATE_CHANGELOG_STORAGE_FACTORIES.clear();
+            Iterator<StateChangelogStorageFactory> iterator =
+                    pluginManager == null
+                            ? ServiceLoader.load(StateChangelogStorageFactory.class).iterator()
+                            : concat(
+                                    pluginManager.load(StateChangelogStorageFactory.class),
+                                    ServiceLoader.load(StateChangelogStorageFactory.class)
+                                            .iterator());
+            iterator.forEachRemaining(
+                    factory ->
+                            STATE_CHANGELOG_STORAGE_FACTORIES.putIfAbsent(
+                                    factory.getIdentifier(), factory));
+            LOG.info(
+                    "StateChangelogStorageLoader initialized with shortcut names {{}}.",
+                    String.join(",", STATE_CHANGELOG_STORAGE_FACTORIES.keySet()));
+        } finally {
+            LOCK.unlock();
+        }
     }
 
     @SuppressWarnings({"rawtypes"})
-    public Iterator<StateChangelogStorage> load() {
-        return concat(
-                pluginManager.load(StateChangelogStorage.class),
-                ServiceLoader.load(StateChangelogStorage.class).iterator());
+    public static StateChangelogStorage load(Configuration configuration) {
+        final String identifier =
+                configuration.getString(CheckpointingOptions.STATE_CHANGE_LOG_STORAGE);
+        LOCK.lock();
```

Comment:
I wonder shall we must have the lock here?

Reply:
Currently we don't need this lock.
I will remove it. Thanks

Reply:
Sorry for bringing this up again, but I think it's actually better to **have** this lock :)
The code is correct without it, but there is an implicit contract that `initialize()` and `load()` are called from the same thread, which is a bit fragile. WDYT?

Reply:
I'm open about this.
Actually, the ```initialize()``` will be called in the early beginning of a taskmanager (**before** other threads run), and this will not change in near future. Thus, it is safe for other thread to access.
However, I admit that this is a strong contract and not good for maintainance.
WDYT? @Myasuka

Reply:
@rkhachatryan , I wonder what the lock will be used if `initialize()` and `load()` are not called from the same thread? For the purpose of preventing concurrent calling `load()` during `initialize()`? However, if the concurrence could happen why we not ensure to call `initialize()` first and then `load()`? If so, maybe we need to add another flag to tell whether `initialized` and print warning or throw exception if `load()` first.

Reply:
I agree, I'd even throw an exception if `load` is called without `initialize`; or if `initialize` called twice.

Reply:
If we really want to make the code stable (although I don't think this problem existed currently), I think we should at least make the code logic clean instead of just adding a lock here.

Reply:
Initializing twice may happen when JVM reuse (eg. Hadoop or unit tests). So I prefer not to change here.